cdap-watchdog/src/main/java/co/cask/cdap/logging/write/LogCleanup.java

```diff
@@ -109,13 +122,167 @@ public Void call() throws Exception {
     }
   }
 
+  /**
+   * Clean log files which does not have corresponding meta data
+   * @param tillTime time till the meta data will be deleted.
+   * @param namespacedLogBaseDirMap namespace to directory map
+   * @param parentDirs parent directories for deleted files
+   */
+  @VisibleForTesting
+  void cleanFilesWithoutMeta(final long tillTime, final Map<String, NamespaceId> namespacedLogBaseDirMap,
+                             final SetMultimap<String, Location> parentDirs) throws Exception {
+    LOG.info("Starting deletion of log files older than {} without metadata", tillTime);
+    List<NamespaceMeta> namespaces = namespaceQueryAdmin.list();
+    // For all the namespaces present in NamespaceMeta, gather log files from disk and log meta. Then delete the log
+    // files for which metadata is not present
+    for (NamespaceMeta namespaceMeta : namespaces) {
+      final NamespaceId namespaceId = namespaceMeta.getNamespaceId();
+      final Location namespacedLogBaseDir = LoggingContextHelper.getNamespacedBaseDirLocation(namespacedLocationFactory,
+                                                                                              logBaseDir, namespaceId,
+                                                                                              impersonator);
+      // processor to get list of log files older than retention duration
+      final Processor<LocationStatus, Set<Location>> diskFilesProcessor = getDiskFilesProcessor(namespaceId, tillTime);
+      // gather all the disk log files older than tillTime for a given namespace
+      if (namespacedLogBaseDir.exists()) {
+        try {
+          impersonator.doAs(namespaceId, new Callable<Void>() {
+            @Override
+            public Void call() throws Exception {
+              Locations.processLocations(namespacedLogBaseDir, true, diskFilesProcessor);
+              return null;
+            }
+          });
+        } catch (Exception e) {
+          LOG.warn("Got exception while accessing path {}", namespacedLogBaseDir, e);
+        }
+        Set<Location> diskFileLocations = diskFilesProcessor.getResult();
+
+        // get the logging context for a given namespace
+        LoggingContext loggingContext = new LogNamespaceLoggingContext(namespaceId.getNamespace());
+
+        // get all the log meta files for a given namespace
+        FileMetaDataManager.TableKey nextTableKey = new FileMetaDataManager.TableKey(loggingContext.getLogPartition(),
+                                                                                     null);
+        Processor<URI, Set<URI>> metaProcessor = getMetaFilesProcessor();
+        // Get all the log metadata in batches of 100
+        do {
+          nextTableKey = fileMetaDataManager.scanFiles(nextTableKey, 100, metaProcessor);
```

Comment:
Good point. It would be good to not store all the files for a namespace in memory at one time though. How about changing the algorithm to remove entries from `diskFileLocations` that are present in meta table? After the complete scan of file meta table is done, we can just delete the remaining entries in `diskFileLocations`.

In addition, we can also limit the number of files in `diskFileLocations`. We can just use an arbitrary limit say 50,000 and stop processing of more files after that for a namespace.